### PR TITLE
Add metrics about LIST handling

### PR DESCRIPTION
**Pull request by MikeSpreitzer**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:
This PR adds metrics about how LIST requests were handled.
Some LIST requests take a lot more CPU per unit of execution duration than most requests, making concurrency limiters (the max-in-flight filter, the API Priority and Fairness (APF) filter) not very effective at managing the CPU usage of apiservers.
We are modifying the APF filter to recognize that some LIST requests are exceptionally costly, but:

1. we need data to tune this, and
2. we want to be able to get useful data from customers in the field.

In a simple run with `hack/local-up-cluster.sh`, this PR contributed 459 lines out of a total of 12340 in a scrape of the kube-apiserver's `/metrics`.

#### Which issue(s) this PR fixes:

Fixes #

#### Special notes for your reviewer:
This is an alternative to #104587 and was originally built on #104652 .  That is, this PR had two commits; the first one is the entirety of #104652 and the second one is what this PR is really about.  Now that #104652 has merged, this PR contains just the second commit.

#### Does this PR introduce a user-facing change?

```release-note
The kube-apiserver's Prometheus metrics have been extended with some that describe the costs of handling LIST requests.  They are as follows.
- *apiserver_list_cache_count*: Counter of LIST requests served from watch cache, broken down by path_prefix, index_used, and predicate_complexity
- *apiserver_list_cache_num_fetched*: Counter of objects read from watch cache in the course of serving a LIST request, broken down by path_prefix and index_used
- *apiserver_list_cache_num_selector_evals*: Counter of objects tested in the course of serving a LIST request from watch cache, broken down by path_prefix and predicate_complexity
- *apiserver_list_cache_num_returned*: Counter of objects returned for a LIST request from watch cache, broken down by path_prefix
- *apiserver_list_etcd3_count*: Counter of LIST requests served from etcd, broken down by resource and predicate complexity
- *apiserver_list_etcd3_queries*: Counter of etcd range queries used to satisfy a LIST request, broken down by resource
- *apiserver_list_etcd3_num_fetched*: Counter of objects read from etcd in the course of serving a LIST request, broken down by resource
- *apiserver_list_etcd3_num_selector_evals*: Counter of objects tested in the course of serving a LIST request from etcd, broken down by resource and predicate complexity
- *apiserver_list_etcd3_num_returned*: Counter of objects returned for a LIST request from etcd, broken down by resource
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by MikeSpreitzer** [comment]
@kubernetes/sig-api-machinery-misc 
/cc @wojtek-t 
/cc @tkashem

**Comment by MikeSpreitzer** [comment]
@deads2k 
@caesarxuchao

**Comment by MikeSpreitzer** [comment]
The force-push to 7cc36474e5d restores some overlooked blank lines in the boilerplate header of `staging/src/k8s.io/apiserver/pkg/storage/cacher/metrics/metrics.go`.

**Comment by wojtek-t** [comment]
@MikeSpreitzer - can you please rebase this PR now?

**Comment by MikeSpreitzer** [comment]
The force-push to fd8f8957219 is the rebase onto `master`.

**Comment by wojtek-t** [comment]
/assign @logicalhan 

@logicalhan - can you please review it from the metrics convention perspective?

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/metrics/metrics.go [hunk 1]
```diff
@@ -85,6 +86,46 @@ var (
 		},
 		[]string{},
 	)
+	listEtcd3Count = compbasemetrics.NewCounterVec(
+		&compbasemetrics.CounterOpts{
+			Name:           "apiserver_list_etcd3_count",
```
Comment (CatherineF-dev) [inline staging/src/k8s.io/apiserver/pkg/storage/etcd3/metrics/metrics.go:91]: `apiserver_list_etcd3_count` -> `apiserver_list_etcd_count`
1. For consistency in metrics.go
2. We only have etcd3 in https://github.com/kubernetes/kubernetes/tree/master/staging/src/k8s.io/apiserver/pkg/storage so far. So using ectd is not ambiguous currently.

Comment (logicalhan) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/metrics/metrics.go:91]: we should use 'storage' instead of 'etcd3', we're actually trying to be resilient with these metrics in the (unlikely) case that we one day swap storage backends.

Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/metrics/metrics.go]: Revised to use "storage", even for the metric that is now "Number of storage range queries used to satisfy a LIST request", in the force-push to 7e956ce0f29 .

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/metrics/metrics.go [hunk 2]
```diff
@@ -85,6 +86,46 @@ var (
...
+		&compbasemetrics.CounterOpts{
+			Name:           "apiserver_list_storage_queries",
+			Help:           "Number of storage range queries used to satisfy a LIST request, split by resource",
+			StabilityLevel: compbasemetrics.ALPHA,
+		},
+		[]string{"resource"},
+	)
+	listStorageNumFetched = compbasemetrics.NewCounterVec(
+		&compbasemetrics.CounterOpts{
+			Name:           "apiserver_list_storage_num_fetched",
```
Comment (logicalhan) [inline staging/src/k8s.io/apiserver/pkg/storage/etcd3/metrics/metrics.go:107]: this is a weird metric... why would we want the cumulative count of objects we've read from lists. Like if I have 200 objects for resource A, and I list three times, this should store 600, but what good is this?

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/metrics/metrics.go [hunk 3]
```diff
@@ -85,6 +86,46 @@ var (
...
+		&compbasemetrics.CounterOpts{
+			Name:           "apiserver_list_storage_count",
+			Help:           "Number of LIST requests served from storage, split by resource and predicate complexity",
+			StabilityLevel: compbasemetrics.ALPHA,
+		},
+		[]string{"resource", "predicate_complexity"},
+	)
+	listStorageQueries = compbasemetrics.NewCounterVec(
+		&compbasemetrics.CounterOpts{
+			Name:           "apiserver_list_storage_queries",
```
Comment (logicalhan) [inline staging/src/k8s.io/apiserver/pkg/storage/etcd3/metrics/metrics.go:99]: this is very strange... this is a count of the paginated requests, right? why are we accumulating the pages per resource?

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/metrics/metrics.go [hunk 4]
```diff
@@ -85,6 +86,46 @@ var (
...
+		&compbasemetrics.CounterOpts{
+			Name:           "apiserver_list_storage_num_fetched",
+			Help:           "Number of objects read from storage in the course of serving a LIST request, split by resource",
+			StabilityLevel: compbasemetrics.ALPHA,
+		},
+		[]string{"resource"},
+	)
+	listStorageNumSelectorEvals = compbasemetrics.NewCounterVec(
+		&compbasemetrics.CounterOpts{
+			Name:           "apiserver_list_storage_num_selector_evals",
```
Comment (logicalhan) [inline staging/src/k8s.io/apiserver/pkg/storage/etcd3/metrics/metrics.go:115]: same idea here...

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/metrics/metrics.go [hunk 5]
```diff
@@ -85,6 +86,46 @@ var (
...
+			Help:           "Number of objects tested in the course of serving a LIST request from storage, split by resource and predicate complexity",
+			StabilityLevel: compbasemetrics.ALPHA,
+		},
+		[]string{"resource", "predicate_complexity"},
+	)
+	listStorageNumReturned = compbasemetrics.NewCounterVec(
+		&compbasemetrics.CounterOpts{
```
Comment (logicalhan) [inline staging/src/k8s.io/apiserver/pkg/storage/etcd3/metrics/metrics.go:122]: okay, so you want to know the diff of objects fetched and returned, it feel like.. but this is cumulative, so i'm not sure how useful this is going to end up being...

**Comment by MikeSpreitzer** [comment]
@logicalhan : all the comments in your last review are really about the same thing.  Let me try to explain.  I first tried to use histograms instead of counts. See #104587 for that.  It increased the volume of an apiserver's metrics by about 33%.  Reviewers judged that to be too much and recommended using counters instead.  I agree that counters will give less detailed information.  However:
- counters give as much information as histograms in a synthetic test scenario with a controlled traffic mix;
- in general usage, the load on an apiserver comes from many requests and the rates at which the counters advance quantify that load;
- part of what we seek to quantify is the relative costs of the various operations involved in handling a LIST request - these operations will have scenario-dependent costs, so having metrics on the individual cost components will help understanding both in profiling and in analyzing observations from the wild.

**Comment by CatherineF-dev** [comment]
> @logicalhan : all the comments in your last review are really about the same thing. Let me try to explain. I first tried to use histograms instead of counts. See #104587 for that. It increased the volume of an apiserver's metrics by about 33%. Reviewers judged that to be too much and recommended using counters instead. I agree that counters will give less detailed information. However:
> 
> * counters give as much information as histograms in a synthetic test scenario with a controlled traffic mix;
> * in general usage, the load on an apiserver comes from many requests and the rates at which the counters advance quantify that load;
> * part of what we seek to quantify is the relative costs of the various operations involved in handling a LIST request - these operations will have scenario-dependent costs, so having metrics on the individual cost components will help understanding both in profiling and in analyzing observations from the wild.

I have a small question upon it. From my understanding, these metrics are used for LIST request profiling. Will these metrics be used only once? After finding the most costly LIST requests, will these metrics still be used?

**Comment by MikeSpreitzer** [comment]
The cost coefficients are usage-dependents.  There are not one right set of answers.  We can do profiling to get an idea of the range of values.  We also want to be prepared to analyze what happened if someone has an unexpected usage and goes well outside the range of what we expected.

**Comment by wojtek-t** [comment]
> I have a small question upon it. From my understanding, these metrics are used for LIST request profiling. Will these metrics be used only once? After finding the most costly LIST requests, will these metrics still be used?

As Mike wrote above - those aren't one-time metrics - they will be useful eventually too [I can easily imagine them being input to auto-tuning once we have that eventually too]

**Comment by CatherineF-dev** [comment]
> > I have a small question upon it. From my understanding, these metrics are used for LIST request profiling. Will these metrics be used only once? After finding the most costly LIST requests, will these metrics still be used?
> 
> As Mike wrote above - those aren't one-time metrics - they will be useful eventually too [I can easily imagine them being input to auto-tuning once we have that eventually too]

Thanks for clarifications!

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go [hunk 6]
```diff
@@ -744,9 +750,21 @@ func (c *Cacher) List(ctx context.Context, key string, opts storage.ListOptions,
 			return err
 		}
 	}
+	metrics.RecordListCacheMetrics(c.resourcePrefix, indexUsed, len(objs), predicateComplexity(pred), numEvals, listVal.Len())
 	return nil
 }
 
+func predicateComplexity(pred storage.SelectionPredicate) int {
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:757]: Going back to the discussion - I'm still not convinced we want that.

I think that knowing if selector was empty vs non-empty is all we need - we won't be able to get any useful information from knowing which exact selector.

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/metrics/metrics.go (hunk 3 above)
Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/metrics/metrics.go:99]: This one I questioned previously too.
In theory, there is a constant cost associated with each etcd query. But having too much data points is making it harder to reason about things.

So I'm personally against introducing this metric.

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/metrics/metrics.go [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+ * ALPHA stability level https://github.com/kubernetes/enhancements/blob/master/keps/sig-instrumentation/1209-metrics-stability/kubernetes-control-plane-metrics-stability.md#stability-classes)
+ *
+ * Promoting the stability level of the metric is a responsibility of the component owner, since it
+ * involves explicitly acknowledging support for the metric across multiple releases, in accordance with
+ * the metric stability policy.
+ */
+var (
+	listCacheCount = compbasemetrics.NewCounterVec(
+		&compbasemetrics.CounterOpts{
+			Name:           "apiserver_list_cache_count",
```
Comment (dgrisonnet) [inline staging/src/k8s.io/apiserver/pkg/storage/cacher/metrics/metrics.go:38]: Following Prometheus best practices, all counter metrics should be suffixed by `_total`. Also, the first part of the metrics name should contain the scope of the metric. So I would recommend renaming this metric to:
```suggestion
			Name:           "apiserver_cache_list_total",
```

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/metrics/metrics.go [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+			Help:           "Number of LIST requests served from watch cache, split by path_prefix, index_used, and predicate_complexity",
+			StabilityLevel: compbasemetrics.ALPHA,
+		},
+		[]string{"path_prefix", "index_used", "predicate_complexity"},
+	)
+	listCacheNumFetched = compbasemetrics.NewCounterVec(
+		&compbasemetrics.CounterOpts{
+			Name:           "apiserver_list_cache_num_fetched",
```
Comment (dgrisonnet) [inline staging/src/k8s.io/apiserver/pkg/storage/cacher/metrics/metrics.go:46]:
```suggestion
			Name:           "apiserver_cache_list_fetched_objects_total",
```
In my suggestion, I removed `num` since the total number is now described by `total` and I added a specification that this metric refers to cache object fetched based on the Help text that describes it. Feel free to change it if you think that my assumption is incorrect.

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/metrics/metrics.go [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+			Help:           "Number of objects read from watch cache in the course of serving a LIST request, split by path_prefix and index_used",
+			StabilityLevel: compbasemetrics.ALPHA,
+		},
+		[]string{"path_prefix", "index_used"},
+	)
+	listCacheNumEvals = compbasemetrics.NewCounterVec(
+		&compbasemetrics.CounterOpts{
+			Name:           "apiserver_list_cache_num_selector_evals",
```
Comment (dgrisonnet) [inline staging/src/k8s.io/apiserver/pkg/storage/cacher/metrics/metrics.go:54]: Based on the help text I would rename this metric:
```suggestion
			Name:           "apiserver_cache_list_evaluated_objects_total",
```

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/metrics/metrics.go [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+ * the metric stability policy.
+ */
+var (
+	listCacheCount = compbasemetrics.NewCounterVec(
+		&compbasemetrics.CounterOpts{
+			Name:           "apiserver_list_cache_count",
+			Help:           "Number of LIST requests served from watch cache, split by path_prefix, index_used, and predicate_complexity",
+			StabilityLevel: compbasemetrics.ALPHA,
+		},
+		[]string{"path_prefix", "index_used", "predicate_complexity"},
```
Comment (dgrisonnet) [inline staging/src/k8s.io/apiserver/pkg/storage/cacher/metrics/metrics.go:42]: According to the data stored in the cache, I would expect the `path_prefix` label to be `resource_prefix`. Also, the `used` part of `index_used` seems to be superflux.

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/metrics/metrics.go [hunk 11]
```diff
@@ -139,3 +185,13 @@ func UpdateLeaseObjectCount(count int64) {
 	// See pkg/storage/etcd3/lease_manager.go
 	etcdLeaseObjectCounts.WithLabelValues().Observe(float64(count))
 }
+
+// RecordListEtcd3Metrics notes various metrics of the cost to serve a LIST request
+func RecordListEtcd3Metrics(resource string, numQueries, numFetched, predicateComplexity, numEvald, numReturned int) {
```
Comment (dgrisonnet) [inline staging/src/k8s.io/apiserver/pkg/storage/etcd3/metrics/metrics.go:190]: Since the metrics were renamed to storage metrics, it would be great to also update the name is the function to align with the other changes.

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go [hunk 12]
```diff
@@ -83,11 +84,11 @@ type objState struct {
 }
 
 // New returns an etcd3 implementation of storage.Interface.
-func New(c *clientv3.Client, codec runtime.Codec, newFunc func() runtime.Object, prefix string, transformer value.Transformer, pagingEnabled bool, leaseManagerConfig LeaseManagerConfig) storage.Interface {
-	return newStore(c, codec, newFunc, prefix, transformer, pagingEnabled, leaseManagerConfig)
+func New(c *clientv3.Client, codec runtime.Codec, newFunc func() runtime.Object, prefix, groupResource string, transformer value.Transformer, pagingEnabled bool, leaseManagerConfig LeaseManagerConfig) storage.Interface {
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go:87]: nit: I suggest moving this parameter plumbing to a separate PR (which we can effectively merge now). And let this PR be purely about metrics.

**Review by jan--f — CHANGES_REQUESTED** [review]
I agree with @dgrisonnet naming suggestions. If the contested storage metrics are kept, they should also be renamed according to this [change request](https://github.com/kubernetes/kubernetes/pull/104723/files#r706139596).

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/metrics/metrics.go [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+	registerMetrics.Do(func() {
+		legacyregistry.MustRegister(listCacheCount)
+		legacyregistry.MustRegister(listCacheNumFetched)
+		legacyregistry.MustRegister(listCacheNumEvals)
+		legacyregistry.MustRegister(listCacheNumReturned)
+	})
+}
+
+// RecordListCacheMetrics notes various metrics of the cost to serve a LIST request
+func RecordListCacheMetrics(pathPrefix, indexUsed string, numFetched, predicateComplexity, numEvals, numReturned int) {
```
[the remark this answers is not in the document]
Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/metrics/metrics.go:84]: In this revision of the PR, this function is called on line 753 in staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go; that's in the List method of Cacher.

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/metrics/metrics.go (hunk 10 above)
Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/metrics/metrics.go:42]: Oh, right, incomplete editing regarding path_prefix vs resource_prefix.

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/metrics/OWNERS [hunk 14]
```diff
@@ -0,0 +1,4 @@
+# See the OWNERS docs at https://go.k8s.io/owners
+
+approvers:
```
[the remark this answers is not in the document]
Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/metrics/OWNERS:3]: This was copied from the etcd3/metrics/OWNERS file.

**Comment by MikeSpreitzer** [comment]
See #104983

**Comment by wojtek-t** [comment]
I like the other one much more - so I'm closing this in favor of #104983

I reassigned reviewers/approvers from here.
